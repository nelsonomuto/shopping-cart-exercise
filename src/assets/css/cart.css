.cart-list {
  margin-bottom: 16px;
}

.cart-item {
  display: inline-grid;
  grid-template-columns: auto 70% 60px;
  grid-template-rows: auto;
  column-gap: 10px;
  row-gap: 1em;
  border-bottom: 1px solid #eee;
  width: 100%;
  padding: 16px 0;
}

.cart-item .cart-item-image {
  align-self: center;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
}
.cart-item .cart-item-image > img {
  margin: 0;
}

.cart-item .cart-item-controls {
  display: flex;
  align-items: center;
  justify-content: center;
}

.cart-item-title {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

.cart-item-price {
  font-size: 0.9em;
  font-weight: bold;
}

.cart-item-controls .cart-item-update-qty-button {
  width: 20px;
  height: 20px;
  border-radius: 50%;
  outline: 0;
  border: 0;
  padding: 0;
  color: rgb(156, 156, 156);
  line-height: normal;
  text-align: center;
  font-size: 0.9em;
  font-weight: bold;
}
.cart-item-controls .cart-item-update-qty-button:hover {
  background: #eee;
  color: #222;
}
.cart-item-controls .cart-item-qty {
  margin: 4px;
  font-size: 0.9em;
}

.cart-portal-container {
  height: 100%;
  width: 100%;
  display: flex;
  position: fixed;
  top: 0;
  left: 0;
  transition: all 0.3s ease-in-out;
}

.cart-portal-container.cart-animate-open {
  background: rgba(0, 0, 0, 0.6);
}

.cart-portal-container .cart-sidebar-container {
  width: 30em;
  background: white;
  transition: all 0.3s cubic-bezier(0.445, 0.05, 0.55, 0.95);
  height: 100vh;
  position: absolute;
  right: 0;
  top: 0;
  transform: translateX(100%);
}

.cart-portal-container.cart-animate-open .cart-sidebar-container {
  opacity: 1;
  transform: translateX(0);
}

.cart-portal-container.cart-animate-closed .cart-sidebar-container {
  opacity: 0;
  transform: translateX(100%);
  pointer-events: auto;
}

.cart-portal-container.cart-animate-closed {
  background: rgba(0, 0, 0, 0);
  pointer-events: none;
}

.cart-sidebar-container {
  display: flex;
  position: relative;
  flex-flow: column nowrap;
  align-items: stretch;
  justify-content: flex-start;
  max-height: 100vh;
  overflow: hidden;
}

.cart-portal-close-button {
  position: absolute;
  top: 10px;
  right: 10px;
  outline: 0;
  border: 0;
}

.cart-portal-close-button::after {
  width: 35px;
  height: 35px;
  content: "\D7";
  border-radius: 50%;
  background: #e4e4e4;
  display: inline-block;
  line-height: 1em;
  font-size: 30px;
  cursor: pointer;
}
.cart-portal-close-button:hover::after {
  background: #efefef;
  color: rgb(71, 171, 230);
}

.cart-sidebar-items-container {
  overflow: scroll;
  padding: 16px;
}
.cart-sidebar-summary-container {
  width: 100%;
  flex: 1 1 auto;
  padding: 16px;
  background: #ddd;
}
.cart-checkout-button {
  font-size: 1.2em;
  font-weight: bold;
  width: 100%;
  padding: 20px;
  background: rgb(0, 190, 73);
  display: flex;
  align-items: center;
  justify-content: center;
  text-transform: uppercase;
  color: #333;
}
.cart-checkout-button:hover {
  background: rgb(74, 251, 141);
  color: rgb(87, 87, 87);
}

.cart-item-price-row {
  vertical-align: text-bottom;
}
.cart-item-remove-button {
  color: rgb(11, 144, 221);
  cursor: pointer;
  outline: 0;
  border: 0;
  margin: 0 6px;
  text-decoration: none;
}
.cart-item-remove-button:hover {
  color: rgb(71, 171, 230);
  text-decoration: underline;
}

.cart-sidebar-summary-container .cart-subtotal {
  margin: 10px 0;
  text-align: right;
}
.cart-sidebar-summary-container .cart-subtotal .cart-subtotal-label {
  font-weight: bold;
  font-size: 0.8em;
  color: #999;
  text-transform: uppercase;
  line-height: 1em;
}
.cart-sidebar-summary-container .cart-subtotal .cart-subtotal-amt {
  font-weight: bold;
  font-size: 1.2em;
  color: #222;
}

.cart-summary {
  text-align: right;
  margin: 20px auto;
}

.cart-summary .cart-summary-total {
  font-weight: bold;
}
