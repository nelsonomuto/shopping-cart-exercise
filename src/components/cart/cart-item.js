import React from "react"

const CartItem = ({ item, onRemove, onIncrement, onDecrement }) => (
  <div>
    <h4 className="cart-item-title">{item.title}</h4>
    <div className="cart-item-controls">
      <button className="cart-item-update-qty-button" onClick={onDecrement}>
        -
      </button>
      <span className="cart-item-qty">{item.qty}</span>
      <button className="cart-item-update-qty-button" onClick={onIncrement}>
        +
      </button>
      <button className="cart-item-remove-button" onClick={onRemove}>
        Remove Item
      </button>
    </div>
  </div>
)

export default CartItem
