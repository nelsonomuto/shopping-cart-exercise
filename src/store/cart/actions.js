const actions = {
  INCREMENT_QUANTITY: "INCREMENT_QUANTITY",
  DECREMENT_QUANTITY: "DECREMENT_QUANTITY",
  ADD_PRODUCT_TO_CART: "ADD_PRODUCT_TO_CART",
  REMOVE_PRODUCT_FROM_CART: "REMOVE_PRODUCT_FROM_CART",
  CLEAR_CART: "CLEAR_CART",
  TOGGLE_CART_OPEN: "TOGGLE_CART_OPEN",
}

export const incrementQuantity = product => ({
  type: actions.INCREMENT_QUANTITY,
  product,
})

export const decrementQuantity = product => ({
  type: actions.DECREMENT_QUANTITY,
  product,
})

export const addProductToCart = product => ({
  type: actions.ADD_PRODUCT_TO_CART,
  product,
})

export const removeProductFromCart = product => ({
  type: actions.REMOVE_PRODUCT_FROM_CART,
  product,
})

export const clearCart = () => ({
  type: actions.CLEAR_CART,
})

export const toggleCartOpen = () => ({
  type: actions.TOGGLE_CART_OPEN,
})

export default actions
