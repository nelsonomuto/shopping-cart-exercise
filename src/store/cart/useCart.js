import React, { createContext, useContext, useEffect, useReducer } from "react"
import store from "./persistentStore"
import reducer, { getCartSummary } from "./reducer"

// Create a "global" context to store our redux state in
const CartStoreContext = createContext()

// Use this to wrap the whole portion of the app that needs access to cart
export const CartStoreProvider = ({ children }) => {
  const initialState = { items: store.get() }
  const [state, dispatch] = useReducer(reducer, initialState)
  return (
    <CartStoreContext.Provider value={{ state, dispatch }}>
      {children}
    </CartStoreContext.Provider>
  )
}

// Use this to access cart state and dispatch functions, like useReducer
// except it is available locally anywhere under the Provider wrapper
export const useCart = () => {
  const { state, dispatch } = useContext(CartStoreContext)

  useEffect(() => {
    store.set(state.items)
  }, [state])

  const nextState = { ...state, summary: getCartSummary(state.items) }
  console.log({ nextState })

  return [nextState, dispatch]
}
