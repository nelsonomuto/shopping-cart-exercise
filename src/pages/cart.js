import React from "react"
import "../assets/css/cart.css"
import CartList from "../components/cart/cart-list"
import CartSummary from "../components/cart/cart-summary"
import Layout from "../components/layout"
import SEO from "../components/seo"

const Cart = () => {
  return (
    <Layout>
      <SEO title="Cart" />
      <h1>Cart</h1>
      <CartList />
      <CartSummary />
    </Layout>
  )
}

export default Cart
