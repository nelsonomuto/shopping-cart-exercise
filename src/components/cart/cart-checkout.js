import React from "react"
import { dollars } from "../../services/formatString"
import { useCart } from "../../store/cart/useCart"
import { clearCart } from "../../store/cart/actions"
import ConfirmationButton from "../confirmation-button"
import CartList from "./cart-list"
import CartSummary from "./cart-summary"

const CartCheckout = () => {
  const [{ items, summary }, dispatch] = useCart()
  return (
    <>
      <CartList />
      <CartSummary />
      {items.length > 0 ? (
        <ConfirmationButton
          className="confirm-order-button"
          confirmText={`Confirm order for ${dollars(summary.totalCost)}`}
          cancelText="Cancel"
          onConfirm={() => {
            alert("Your order has been confirmed!")
            dispatch(clearCart())
          }}
        >
          Place Order
        </ConfirmationButton>
      ) : null}
    </>
  )
}

export default CartCheckout
