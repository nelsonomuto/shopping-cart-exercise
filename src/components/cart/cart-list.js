import { Link } from "gatsby"
import React from "react"
import { clearCart } from "../../store/cart/actions"
import { useCart } from "../../store/cart/useCart"
import CartItem from "./cart-item"

const CartList = ({ onShop }) => {
  const [{ items }, dispatch] = useCart()

  if (items.length === 0) {
    return (
      <div className="cart-list cart-list-empty">
        <h3>
          Your cart is empty.{" "}
          <Link to="/" onClick={onShop}>
            Shop now
          </Link>
        </h3>
      </div>
    )
  }

  return (
    <div className="cart-list-container">
      <div className="cart-list">
        {items.map(item => (
          <CartItem item={item} key={item.sku} />
        ))}
      </div>
      <button
        className="cart-list-clear-button"
        onClick={() => dispatch(clearCart())}
      >
        Clear Cart
      </button>
    </div>
  )
}

export default CartList
