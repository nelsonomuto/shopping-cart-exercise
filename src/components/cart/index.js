import { Link } from "gatsby"
import React, { useEffect, useRef } from "react"
import { dollars } from "../../services/formatString"
import { toggleCartOpen } from "../../store/cart/actions"
import { useCart } from "../../store/cart/useCart"
import CartList from "./cart-list"
import CartPortal from "./cart-portal"

const Cart = () => {
  const [{ isCartOpen, summary }, dispatch] = useCart()
  const containerRef = useRef(null)

  const toggleCart = () => dispatch(toggleCartOpen())

  const onClickOutside = e => {
    if (containerRef.current.contains(e.target)) {
      // inside
      return
    }
    // outside
    toggleCart()
  }

  useEffect(() => {
    if (isCartOpen) {
      document.addEventListener("mousedown", onClickOutside)
      return () => {
        document.removeEventListener("mousedown", onClickOutside)
      }
    } else {
      document.removeEventListener("mousedown", onClickOutside)
    }
  }, [isCartOpen])

  return (
    <CartPortal>
      <div
        onClick={onClickOutside}
        className={
          "cart-portal-container" +
          (isCartOpen ? " cart-animate-open" : " cart-animate-closed")
        }
      >
        <div className="cart-sidebar-container" ref={containerRef}>
          <button className="cart-portal-close-button" onClick={toggleCart}>
            Close
          </button>
          <div className="cart-sidebar-items-container">
            <h1>Cart</h1>
            <CartList onShop={toggleCart} />
          </div>
          <div className="cart-sidebar-summary-container">
            <div className="cart-subtotal">
              Subtotal: {dollars(summary.subtotal)}
            </div>
            <Link to="/checkout" className="cart-checkout-button">
              Checkout
            </Link>
          </div>
        </div>
      </div>
    </CartPortal>
  )
}

export default Cart
