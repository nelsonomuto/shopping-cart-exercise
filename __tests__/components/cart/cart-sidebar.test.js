import { mount, shallow } from "enzyme"
import React from "react"
import CartSidebar from "../../../src/components/cart/cart-sidebar"
import { useCart } from "../../../src/store/cart/useCart"
import { toggleCartOpen } from "../../../src/store/cart/actions"
import { act } from "react-dom/test-utils"

const { getDefaultState } = jest.requireActual(
  "../../../src/store/cart/useCart"
)

const mockProduct = {
  sku: "fake-sku",
  title: "fake-title",
  price: 999.99,
  image: "fake-image-url",
}

const mockItem = {
  ...mockProduct,
  qty: 1,
}

const mockState = getDefaultState({ items: [mockItem], isCartOpen: true })

// Mock useCart which is required in the component.
// Don't test that side effects from useCart actually happen,
// just test that dispatch is called where needed and that
// the original state gets rendered.
const mockDispatch = jest.fn()
jest.mock("../../../src/store/cart/useCart.js", () => ({
  useCart: jest.fn(),
}))

describe("<CartSidebar />", () => {
  beforeEach(() => {
    useCart.mockImplementation(() => [mockState, mockDispatch])
  })

  afterEach(() => {
    document.addEventListener = jest.fn()
  })

  it("matches snapshot", () => {
    const component = shallow(<CartSidebar />)
    expect(component).toMatchSnapshot()
  })
  it("matches snapshot (not open)", () => {
    useCart.mockImplementation(() => [
      getDefaultState({ items: [mockItem], isCartOpen: false }),
      mockDispatch,
    ])
    const component = shallow(<CartSidebar />)
    expect(component).toMatchSnapshot()
  })

  it("close button closes cart", () => {
    const component = mount(<CartSidebar />)
    const closeButton = component.find(".cart-portal-close-button")
    act(() => {
      closeButton.simulate("click")
    })
    component.update()
    expect(mockDispatch).toHaveBeenCalledWith(toggleCartOpen())
    expect(
      component.find(".cart-portal-container").hasClass(".cart-animate-closed")
    ).toBeTrue
  })

  it("clicking outside closes cart", () => {
    const map = {}
    document.addEventListener.mockImplementationOnce((event, cb) => {
      map[event] = cb
    })
    const component = mount(<CartSidebar />)
    const containerComponent = component.find(".cart-portal-container")
    // simulate click outside
    act(() => {
      map.mousedown({ target: containerComponent.instance() })
    })
    component.update()

    expect(mockDispatch).toHaveBeenCalledWith(toggleCartOpen())
  })

  it("clicking inside doesnt close cart", () => {
    const map = {}
    document.addEventListener.mockImplementationOnce((event, cb) => {
      map[event] = cb
    })
    const component = mount(<CartSidebar />)
    const containerComponent = component.find(".cart-sidebar-container")
    // simulate click outside
    act(() => {
      map.mousedown({ target: containerComponent.instance() })
    })
    component.update()

    expect(mockDispatch).not.toHaveBeenCalledWith(toggleCartOpen())
  })

  it("unmount removes event listener", () => {
    useCart.mockImplementation(() => [
      getDefaultState({ items: [], isCartOpen: true }),
      mockDispatch,
    ])
    const map = {}
    document.removeEventListener = jest.fn()
    document.addEventListener = jest.fn((event, cb) => {
      map[event] = cb
    })

    const component = mount(<CartSidebar />)

    act(() => {
      component.unmount()
    })
    expect(document.removeEventListener).toHaveBeenCalledWith(
      "mousedown",
      map.mousedown
    )
  })
})

describe("<CartSidebar /> (with no items)", () => {
  it("matches snapshot (empty state)", () => {
    useCart.mockImplementationOnce(() => [
      getDefaultState({ items: [], isCartOpen: true }),
      mockDispatch,
    ])
    const component = shallow(<CartSidebar />)
    expect(component).toMatchSnapshot()
  })
})
