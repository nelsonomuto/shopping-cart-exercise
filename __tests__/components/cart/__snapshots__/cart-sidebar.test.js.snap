// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartSidebar /> (with no items) matches snapshot (empty state) 1`] = `
<CartPortal>
  <div
    className="cart-portal-container cart-animate-open"
  >
    <div
      className="cart-sidebar-container"
    >
      <button
        className="cart-portal-close-button"
        onClick={[Function]}
      />
      <div
        className="cart-sidebar-items-container"
      >
        <h1>
          Cart
        </h1>
        <CartList
          onShop={[Function]}
        />
      </div>
      <div
        className="cart-sidebar-summary-container"
      >
        <div
          className="cart-subtotal"
        >
          <div
            className="cart-subtotal-label"
          >
            Subtotal:
          </div>
          <div
            className="cart-subtotal-amt"
          >
            $ 0.00
          </div>
        </div>
      </div>
    </div>
  </div>
</CartPortal>
`;

exports[`<CartSidebar /> matches snapshot (not open) 1`] = `
<CartPortal>
  <div
    className="cart-portal-container cart-animate-closed"
  >
    <div
      className="cart-sidebar-container"
    >
      <button
        className="cart-portal-close-button"
        onClick={[Function]}
      />
      <div
        className="cart-sidebar-items-container"
      >
        <h1>
          Cart
        </h1>
        <CartList
          onShop={[Function]}
        />
      </div>
      <div
        className="cart-sidebar-summary-container"
      >
        <div
          className="cart-subtotal"
        >
          <div
            className="cart-subtotal-label"
          >
            Subtotal:
          </div>
          <div
            className="cart-subtotal-amt"
          >
            $ 999.99
          </div>
        </div>
        <mockConstructor
          className="cart-checkout-button"
          to="/checkout"
        >
          Checkout
        </mockConstructor>
      </div>
    </div>
  </div>
</CartPortal>
`;

exports[`<CartSidebar /> matches snapshot 1`] = `
<CartPortal>
  <div
    className="cart-portal-container cart-animate-open"
  >
    <div
      className="cart-sidebar-container"
    >
      <button
        className="cart-portal-close-button"
        onClick={[Function]}
      />
      <div
        className="cart-sidebar-items-container"
      >
        <h1>
          Cart
        </h1>
        <CartList
          onShop={[Function]}
        />
      </div>
      <div
        className="cart-sidebar-summary-container"
      >
        <div
          className="cart-subtotal"
        >
          <div
            className="cart-subtotal-label"
          >
            Subtotal:
          </div>
          <div
            className="cart-subtotal-amt"
          >
            $ 999.99
          </div>
        </div>
        <mockConstructor
          className="cart-checkout-button"
          to="/checkout"
        >
          Checkout
        </mockConstructor>
      </div>
    </div>
  </div>
</CartPortal>
`;
